pipelines/toast_ground_sim.py

```diff
@@ -547,415 +162,173 @@ def parse_arguments(comm):
-        quat = focalplane[det]["quat"]
-        vec = qa.rotate(quat, ZAXIS)
-        cosangs.append(np.dot(ZAXIS, vec))
-    mincos = np.amin(cosangs)
-    maxdist = max(np.degrees(np.arccos(mincos)), rmin)
-    return maxdist * 1.001
+    return args, comm
 
 
 @function_timer
 def load_focalplanes(args, comm, schedules):
     """ Attach a focalplane to each of the schedules.
 
+    Args:
+        schedules (list) :  List of tuples of the form
+            (`site`, `all_ces`) where `site` is a Site
+            instansce with the `telescope` field filled
+            and `all_ces` is a list of CES objects.
+    Returns:
+        detweights (dict) : Inverse variance noise weights for every
+            detector across all focal planes.
     """
-    tmr = Timer()
-    tmr.start()
+    timer = Timer()
+    timer.start()
 
+    telescopes = []
+    for schedule in schedules:
+        telescopes.append(schedule[0].telescope)
     # Load focalplane information
 
     focalplanes = []
     if comm.world_rank == 0:
-        ftmr = Timer()
-        for fpfile in args.fp.split(","):
-            ftmr.start()
+        ftimer = Timer()
+        for fpfile in args.focalplane.split(","):
+            ftimer.start()
             with open(fpfile, "rb") as picklefile:
                 focalplane = pickle.load(picklefile)
                 focalplanes.append(focalplane)
-                ftmr.report_clear("Load {}".format(fpfile))
-        ftmr.stop()
-    focalplanes = comm.comm_world.bcast(focalplanes)
+                ftimer.report_clear("Load {}".format(fpfile))
+        ftimer.stop()
+    if comm.comm_world is not None:
+        focalplanes = comm.comm_world.bcast(focalplanes)
+        telescopes = comm.comm_world.bcast(telescopes)
 
     if len(focalplanes) == 1 and len(schedules) > 1:
         focalplanes *= len(schedules)
+    if len(telescopes) == 1 and len(schedules) > 1:
+        telescopes *= len(schedules)
     if len(focalplanes) != len(schedules):
         raise RuntimeError(
             "Number of focalplanes must equal number of schedules or be 1."
         )
 
+    # Append a focal plane and telescope to each entry in the schedules list
     detweights = {}
-    for schedule, focalplane in zip(schedules, focalplanes):
+    for schedule, focalplane, telescope in zip(schedules, focalplanes, telescopes):
         schedule.append(focalplane)
-        for detname, det in focalplane.items():
-            net = det["NET"]
-            detweight = 1.0 / (args.samplerate * net * net)
+        schedule.append(Telescope(telescope))
+        for detname, detdata in focalplane.items():
+            net = detdata["NET"]
+            detweight = 1.0 / (args.sample_rate * net * net)
             if detname in detweights and detweights[detname] != detweight:
                 raise RuntimeError("Detector weight for {} changes".format(detname))
             detweights[detname] = detweight
 
-    tmr.stop()
+    timer.stop()
     if comm.world_rank == 0:
```

Comment:
can we have the check of rank 0 be inside `timer`?

Reply:
I suppose we could, but we might also be interested in the timings on individual processes.

Reply:
I meant for printing purposes, it automatically prints only on rank 0, so you can use it without if

Reply:
Should the Timer.report() method then take the process rank as an argument?

Reply:
I was thinking about the constructor instead. How does `Timer` know which rank is it reporting about now?

Reply:
It does not. I would prefer deferring modifications to the Timer class to some other time.

Reply:
Ok

Reply:
Just a comment here about timers- you can also look at the C++ source for more details:

1.  The "Timer" class is a serial object.  If you only want to time events on rank zero, one could just create the timer on that rank.  Standalone "Timer" instances basically replace the functionality of using MPI_Wtime to get elapsed time on one process and print a message.

2.  Additionally each process has a singleton dictionary of timers (GlobalTimers https://github.com/hpc4cmb/toast/blob/refactor_2.3/src/libtoast/include/toast/sys_utils.hpp#L104) which can be started / stopped anywhere in the code.  The @function_timer decorator adds entries to this dictionary.  This global timer dictionary is also serial with no MPI awareness.

3.  When collecting timings across multiple processes, there are functions to do that:  https://github.com/hpc4cmb/toast/blob/refactor_2.3/src/toast/timing.py#L103

4.  Although one could define an "MPITimer" or "SynchronousTimer" class which took a communicator and then did a barrier and a "report" on only a single process, that is not currently implemented.  This would need to be a new additional class, since there are definitely use cases for plain serial timers even if MPI is enabled (for example if different processes are following different code paths and doing different things).  I don't object to defining such a class if you want.  This would be defined in libtoast_mpi with bindings in _libtoast_mpi.cpp.  You could also consider a thin wrapper statement in toast.timing that fell back to using a plain Timer in the case where MPI is disabled.